mlir/lib/Transform/AIRMiscPasses.cpp

```diff
@@ -879,6 +881,119 @@ void AIRLowerHerdParallelPass::runOnOperation() {
   (void)applyPatternsAndFoldGreedily(op, std::move(patterns));
 }
 
+class HoistScfChannelGetPutPass : public RewritePattern {
+public:
+  HoistScfChannelGetPutPass(MLIRContext *context, PatternBenefit benefit = 1)
+      : RewritePattern(MatchAnyOpTypeTag(), benefit, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto chanOp = dyn_cast<air::ChannelInterface>(op);
+    if (!chanOp)
+      return failure();
+    MemRefType memrefTy = chanOp.getMemref().getType().cast<MemRefType>();
+
+    // channel op must be in an scf.for loop
+    auto forOp = op->getParentOfType<scf::ForOp>();
+    if (!forOp)
+      return failure();
+
+    // channel op must be the only op in the loop
+    if (forOp.getBody()->getOperations().size() != 2)
+      return failure();
+
+    // scf.for must have constant bounds and step
+    auto lb = forOp.getLowerBound().getDefiningOp<arith::ConstantIndexOp>();
+    auto ub = forOp.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    auto step = forOp.getStep().getDefiningOp<arith::ConstantIndexOp>();
+    if (!lb || !ub || !step)
```

Comment:
```suggestion
    if (!(lb && ub && step))
```
might be easier to understand the precondition you expect.

Reply:
I prefer the one I can parse left-to-right without a mental stack.

Reply:
I did not think you were a Turing machine! ;-)